index.bs

```diff
@@ -1149,6 +1149,9 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
             Note: Modifying or removing [=list/items=] from the value returned from {{AuthenticatorAttestationResponse/getTransports()}}
             could negatively impact user experience, or even prevent use of the corresponding credential.
 
+        :   <dfn>UV</dfn>
+        ::  A Boolean value indicating whether this [=credential=] has reported any successful [=user verification=].
```

Comment:
Wouldn't it be the authenticator, not the credential?

Reply:
Hmyeah, with the "whether X has reported UV" wording I suppose that would be more correct, but my thinking was that the RP only really sees the credential (and doesn't, for example, know whether a credential with `UV=0` is on the same authenticator as a different credential with `UV=1`).

How about this, then? A bit more verbose, but perhaps not too bad.

```suggestion
        ::  A Boolean value indicating whether any [=credential=] from this [=public key credential source=] has had the [=authData/flags/UV=] [=flag=] set.
```

Reply:
I think the "any successful user verification" is a bit weird, "any" could mean authentication or registration ceremonies. Shouldn't this be "has reported successful userverification during registration".

Reply:
I don't think so, is there a reason you shouldn't update your knowledge about the credential's capabilities in authentication ceremonies too?

Reply:
I'm thinking "out loud" a bit in this response. I'm not sure what's right here. 

I think you've mentioned that when an authenticator changes from UV=0 to UV=1 during it's lifecycle, then it's not assured that it's the same user who registered that. There are really two ways to look at it I guess. Should the UV be enforced if it changes in the lifecycle? In my mind I'm thinking about scenarios like below. This can happen when a user has changed browser say from Firefox (which currently only does CTAP1, but soon! CTAP 2 SOON!) and swaps to chrome which does CTAP2 and now can conduct UV. 

* The authenticator registers with UV=0
* The authenticator authenticate with UV=0
* The authenticator authenticates with UV=1 

1) The user now has setup a pin because they never had the option before.
2) Someone malicious has setup a pin on the token without the persons knowledge
3) worst case, a user has setup a pin and forgotten it.

So by updating the knowledge of UV=1, you prevent the ability to "go back" to firefox after this (which as mentioned should not be a concern in the future ....). It makes it a one way path. 

But the worse case is when this UV=0 goes to UV=1, it's a risk of not being the same user as you have previously mentioned. 

I think to me, I'm thinking about this from the view of "consistency". How would a human think about this? What did the user register, and what experience do they expect to continue? If you register with UV=0, you should never need UV in the future. The is especially true with password + token where the token pin shouldn't really be needed. The UV state is better with a self contained MFA for example, and having that UV state from the very start from registration. Again creating that consistent expectation. 

Sadly CTAP2.0 (?) I think ruins this by enforcing UV=1 on registration under discouraged, but UV=0 during authentication, so this is where it becomes confusing to store UV alone (In webauthn-rs we store UV bool + the policy used at registration to understand this context). But this is very subtle. 

But there is also certainly an argument to be made for "improving" the state of UV on a credential as it goes forward.

So I think I'd err on the site of UV being "this was the state of UV at registration" because I think that is reflective more of the "if UV is requested, we know it was frm the same user who registered the device". Allowing UV to change post credential creation, means that the UV boolean we store is just "Someone has UV'd at some point". 

Does that make sense/help?

Reply:
> > So by updating the knowledge of UV=1, you prevent the ability to "go back" to firefox after this
> 
> Ah, ok, that's a very good point. While yes, that should hopefully be a temporary problem, it doesn't seem quite fair to end-users that we make this recommendation right now.

Agreed, 

> 
> > But the worse case is when this UV=0 goes to UV=1, it's a risk of not being the same user as you have previously mentioned.
> 
> Yes, but only if authenticator possession is the only factor you check - and in that case, you (the RP) have evidently already accepted that risk. And the risk can of course be mitigated by using some other second authentication factor, such as a password, until you've seen UV=1 for the first time.

True, but if you the RP accept UV=0, why would you want to change to UV=1 going forward? Wouldn't you want to re-enroll the credential as a self contained MFA than "upgrading in place" a second factor to be that self contained MFA? 

> 
> > I think to me, I'm thinking about this from the view of "consistency". [...] If you register with UV=0, you should never need UV in the future.
> 
> Agreed, and that will continue to be the case if the RP consistently sets `userVerification: "discouraged"`. This case of "upgrading" to requiring UV is _specifically_ targeted for when the RP sets `userVerification: "preferred"` - because that inherently means the RP is interested in such an upgrade, otherwise they should use `"discouraged"` or `"required"` instead.

well, UV=discouraged sadly with CTAP2.0 is always UV=1 at registration which breaks consistency, but at least in that scenario you know that the if you change from discouraged to required, then you have the same user who verified at registration? 

> 
> > So I think I'd err on the site of UV being "this was the state of UV at registration" because I think that is reflective more of the "if UV is requested, we know it was frm the same user who registered the device". Allowing UV to change post credential creation, means that the UV boolean we store is just "Someone has UV'd at some point".
> 
> That's a fair point, though. Maybe this is better as a separate section of less prescriptive guidance (or even an external resource like [How to FIDO](https://fido-alliance.github.io/how-to-fido/HowToFIDO.html)), because there really isn't a single correct way to handle this in all contexts.

So my thinking here is that the UV bool stored by an RP should be the UV state as it was at registration so that:

* If it's UV=0, they can see that UV=1 may or may not be the same user who registered the device, so the RP can make that risk assesment.
* That if it's UV=1 at registration, and has been discouraged, you can upgrade to required/preferred knowing the same user who registered the device as is now authenticating with it. 
* That UV stored by the RP should be the UV state at registration to encourage "consistency" matching user expectations of the credential and it's lifecycle.
* That we recommend the UV state is NOT updated for the lifecycle of the credential because that gets all kinds of confusing for everyone involved.

It's a more conservative stance, but it can be relaxed in the future if needed - where if we go the other way and recommend UV is updated during auth, and we find a problem or change our mind, it's harder to "constrict" and there will remain RP's that will never fix that behaviour - which again will lead to broader inconsistency in the world of RP's :) 

What do you think?

Reply:
@emlun It seems FF have been delayed updating to ctap2.1, so it may not be til later this year, so I think the "downgrade" capability has to remain for now :(

Reply:
I've made the change 7e541a0c9cedea58006f18a414c778d3c1e4f026 which I believe should address @timcappalli's original concern. The "Firefox downgrade issue" remains, though. I don't know a good solution for that other than waiting for Firefox to release CTAP2 support before merging this.